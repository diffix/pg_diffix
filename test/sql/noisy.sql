LOAD 'pg_diffix';

SET pg_diffix.noise_layer_sd = 7;
SET pg_diffix.low_count_layer_sd = 2;
SET pg_diffix.low_count_min_threshold = 2;

SET ROLE diffix_test;
SET pg_diffix.session_access_level = 'anonymized_trusted';

----------------------------------------------------------------
-- Sanity checks
----------------------------------------------------------------

SELECT diffix.access_level();

----------------------------------------------------------------
-- Basic queries
----------------------------------------------------------------

SELECT COUNT(*) FROM test_customers;
SELECT COUNT(*) FROM test_purchases;

SELECT COUNT(city), COUNT(DISTINCT city) FROM test_customers;

SELECT COUNT(DISTINCT cid) FROM test_purchases;

SELECT city, COUNT(DISTINCT id) FROM test_customers GROUP BY 1;

----------------------------------------------------------------
-- Basic queries - expanding constants in target expressions
----------------------------------------------------------------

SELECT 1 FROM test_patients;
SELECT cast(1 as real) FROM test_patients;
SELECT 1, COUNT(*) FROM test_patients;
SELECT 1, city FROM test_customers;
SELECT city, 'aaaa' FROM test_customers;

----------------------------------------------------------------
-- Multi-AID queries
----------------------------------------------------------------

SELECT city FROM test_patients GROUP BY 1;

SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM test_patients;

----------------------------------------------------------------
-- LCF & Filtering
----------------------------------------------------------------

SELECT id FROM test_customers;

SELECT city FROM test_customers;

SELECT city FROM test_customers GROUP BY 1 HAVING length(city) <> 4;

SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM london_customers;

-- LCF doesn't depend on the bucket seed, both queries should have same noisy threshold.
SELECT diffix.floor_by(age, 30), COUNT(*) FROM test_patients GROUP BY 1;
SELECT diffix.floor_by(age, 30), diffix.floor_by(age, 106), COUNT(*) FROM test_patients GROUP BY 1, 2;

----------------------------------------------------------------
-- Empty tables
----------------------------------------------------------------

SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM empty_test_customers;
