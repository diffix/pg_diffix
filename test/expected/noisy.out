SET pg_diffix.noise_layer_sd = 7;
SET pg_diffix.low_count_layer_sd = 3;
SET pg_diffix.session_access_level = 'publish';
----------------------------------------------------------------
-- Utilities
----------------------------------------------------------------
SELECT diffix.access_level();
 access_level 
--------------
 publish
(1 row)

----------------------------------------------------------------
-- Basic queries
----------------------------------------------------------------
SELECT COUNT(*) FROM test_customers;
 count 
-------
    15
(1 row)

SELECT COUNT(*) FROM test_purchases;
 count 
-------
    23
(1 row)

SELECT COUNT(city), COUNT(DISTINCT city) FROM test_customers;
 count | count 
-------+-------
    14 |     1
(1 row)

SELECT COUNT(DISTINCT cid) FROM test_purchases;
 count 
-------
     9
(1 row)

SELECT city, COUNT(DISTINCT id) FROM test_customers GROUP BY 1;
  city  | count 
--------+-------
 Rome   |    14
 Berlin |    15
(2 rows)

----------------------------------------------------------------
-- Basic queries - expanding constants in target expressions
----------------------------------------------------------------
SELECT 1 FROM test_patients;
 ?column? 
----------
        1
        1
        1
        1
        1
(5 rows)

SELECT cast(1 as real) FROM test_patients;
 float4 
--------
      1
      1
      1
      1
      1
(5 rows)

SELECT 1, COUNT(*) FROM test_patients;
 ?column? | count 
----------+-------
        1 |     5
(1 row)

SELECT 1, city FROM test_customers;
 ?column? |  city  
----------+--------
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Rome
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
        1 | Berlin
(29 rows)

SELECT city, 'aaaa' FROM test_customers;
  city  | ?column? 
--------+----------
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Rome   | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
 Berlin | aaaa
(29 rows)

----------------------------------------------------------------
-- Multi-AID queries
----------------------------------------------------------------
SELECT city FROM test_patients GROUP BY 1;
 city 
------
(0 rows)

SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM test_patients;
 count | count | count 
-------+-------+-------
     5 |     5 |     0
(1 row)

----------------------------------------------------------------
-- LCF & Filtering
----------------------------------------------------------------
SELECT id FROM test_customers;
 id 
----
(0 rows)

SELECT city FROM test_customers;
  city  
--------
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Rome
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
 Berlin
(29 rows)

SELECT city FROM test_customers GROUP BY 1 HAVING length(city) <> 4;
  city  
--------
 Berlin
(1 row)

SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM london_customers;
 count | count | count 
-------+-------+-------
     0 |     0 |     0
(1 row)

----------------------------------------------------------------
-- Empty tables
----------------------------------------------------------------
SELECT COUNT(*), COUNT(city), COUNT(DISTINCT city) FROM empty_test_customers;
 count | count | count 
-------+-------+-------
     0 |     0 |     0
(1 row)

